Add section about kube-state-metrics (#45509)
kubernetes/website: content/en/docs/concepts/cluster-administration/kube-state-metrics.md

---

> title: Metrics for Kubernetes object state

Reviewer: ```suggestion
title: Metrics for Kubernetes Object States
```
(per our style guide)

content_type: concept
weight: 75
description: >-
   kube-state-metrics, an add-on agent to generate and expose cluster-level metrics.
---

The state of Kubernetes objects in the Kubernetes API can be exposed as metrics.
An add-on agent called [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) can connect to the Kubernetes API server and expose a HTTP endpoint with metrics generated from the state of individual objects in the cluster.
It exposes various information about the state of objects like labels and annotations, startup and termination times, status or the phase the object currently is in.
For example, containers running in pods create a `kube_pod_container_info` metric.

> This includes the name of the container, the name of the pod it is part of, the namespace the pod is running in, the name of the container image, the ID of the image, the spec of the image, the ID of the running container and the ID of the pod as labels.

Reviewer: For clarity, try:
```suggestion
This includes the name of the container, the name of the pod it is part of, the
{{< glossary_tooltip text="namespace" term_id="namespace" >}} the pod is running in,
the name of the container image, the ID of the image, the spec of the image, the ID of
the running container and the ID of the pod. These details are exposed as OpenMetrics
labels.
```

Is “the spec of the image” a fragment of JSON? It's not obvious from the text what that label's value might be.

Author: It's the image name from the spec https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/pod.go#L154 and the rest comes from the container status.

> 

Reviewer: Because we're mentioning an external tool (Prometheus, which is not part of Kubernetes), we should add:

```suggestion

{{% thirdparty-content single="true" %}}

```

> An external component that is able and capable to scrape the endpoint of kube-state-metrics (e.g. Prometheus) can now be used to enable the following use cases.

Reviewer: (nit) We [avoid Latin phrases](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases) and their abbreviations.

> # Using Metrics from kube-state-metrics to Query the Cluster State

Reviewer: - Use level 2 (or higher) headings
- In concept pages, typical headings are also concepts
- Please use sentence case here

Try:
```suggestion
## Example: using metrics from kube-state-metrics to query the cluster state {#example-kube-state-metrics-query-1}
```

Metric series generated by kube-state-metrics are helpful to gather further insights into the cluster, as they can be used for querying.

> The following PromQL query returns the number of pods that are not ready:

Reviewer: ```suggestion
If you use Prometheus or another tool that uses the same query language, the following
PromQL query returns the number of pods that are not ready:
```

```
count(kube_pod_status_ready{condition="false"}) by (namespace, pod)
```

> # Using Metrics from kube-state-metrics for Alerting

Reviewer: - Use level 2 (or higher) headings
- In concept pages, typical headings are also concepts

Try:
```suggestion
## Example: alerting based on metrics from kube-state-metrics {#example-kube-state-metrics-alert-1}
```

Metrics generated from kube-state-metrics also allow for alerting on issues in the cluster.

> The following alert will fire if there are pods that have been in `Terminating` state for more than 5 minutes:

Reviewer: ```suggestion
If you use Prometheus, the following alert will fire if there are pods that have been in a `Terminating` state for more 
than 5 minutes:
```
